Cross Validation

###**INSTRUCTIONS**
#####Provide your observation/summary for all tutorials and tasks. Marks will be deducted for those who dont write the summary.
#####Print this file as pdf and submit both pdf and .ipynb file(No screenshots). Not submitting any of these files will impact their grades.

###Tutorial

####Performing EDA on the dataset

In [ ]:
import pandas as pd
import numpy as np
import time
from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score, LeaveOneOut,\
RepeatedKFold, train_test_split
from sklearn.preprocessing import OneHotEncoder, RobustScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.ensemble import RandomForestClassifier

here, we are importing the libraries for performing data manipulation and also machine learning from scikit learn. here we are using some libraries for cross validation and data preprocessing such as onehotencoder, robustscaler, and we are here creating the machine learning model and for that we are using randomforest classifier. here we are using the pipeline to perform the model training an evaluation.

In [ ]:
df = pd.read_csv('income_evaluation-1.csv', na_values=' ?')
df.head()

age          workclass   fnlwgt   education   education-num  \
0   39          State-gov    77516   Bachelors              13   
1   50   Self-emp-not-inc    83311   Bachelors              13   
2   38            Private   215646     HS-grad               9   
3   53            Private   234721        11th               7   
4   28            Private   338409   Bachelors              13   

        marital-status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

    capital-gain   capital-loss   hours-per-week  native-country  income  
0           2174              0               40   United-States   <=50K  
1              0              0               13   United-States   <=50K  
2              0              0               40   United-States   <=50K  
3              0              0               40   United-States   <=50K  
4              0              0               40            Cuba   <=50K

here, we are reading the csv file income_evaluation-1 with pandas library as a dataframe, where ? is work with the  missing values.
here, we are displaying first five records from the dataframe and records columns are age, workclass, education and occupation etc.here this data is based on the income classification with different features.


In [ ]:
df.isna().sum()

age                   0
 workclass         1836
 fnlwgt               0
 education            0
 education-num        0
 marital-status       0
 occupation        1843
 relationship         0
 race                 0
 sex                  0
 capital-gain         0
 capital-loss         0
 hours-per-week       0
 native-country     583
 income               0
dtype: int64

here, we are using df with isna and sum() for counting the number of missing values of the each column. focusing on the counts workclass have 1836 missing values, for occupation its 1843 and for native country its 583 columns. here, we are calculating the missing value with the column name to check which column data is missing most so we can handle that data with removal or replacing it for creating modeling.


In [ ]:
df.fillna('missing', inplace=True)

here, we are using df with fillna for the handling missing values to modefied in the dataframe and using inplace as true. this is the basic way to handing the missing values. here, we are replacing nan values with the useful placeholder.


In [ ]:
df.isna().sum()

age                0
 workclass         0
 fnlwgt            0
 education         0
 education-num     0
 marital-status    0
 occupation        0
 relationship      0
 race              0
 sex               0
 capital-gain      0
 capital-loss      0
 hours-per-week    0
 native-country    0
 income            0
dtype: int64

here, again we are using df with isna() with sum().here, we are counting the columns who have nan values or missing values. focusing on the counting of columns now columns with missing values are count as 0, which suggest that there is no missing values in the dataframe.

In [ ]:
X = df.drop(' income', axis=1)
y = df[' income']

here, we are dropping the column income from the dataframe and using all other columns and and all of them stored in variable x, while in contrast, in the y variable income column is use for model.
we are here using x and y variable  for creating model and fitting the data into model using x and y.

In [ ]:
X.shape

(32561, 14)

here, we are checking the shape of the data for x variable. here we are displaying the x.shape, which is 32561 rows with 14 columns.

In [ ]:
y.value_counts()

 <=50K    24720
 >50K      7841
Name:  income, dtype: int64

here, we are counting the y values using the y.value_counts(). focusing on the results <= 50 k is 24720 and >= 50k  is 7841.

###1. KFold Cross Validation

In [ ]:
kf = KFold(n_splits=5)

cross validation is the resampling method use for the different part of the data for test and train on the different iteration. this is use for set the main goal for prediction, this is the estimate that how our prediction model  performing well and accurate with the data.

k-fold is the technique for evaluation the prediction model. here, we are dividing into the subset of the k or in the folds. here, we are training the model as well as evaluating for the k times,  we are using the different kind of the folds as the validating set for each of the time. here , we are calculating the performance metrics for the each of the folds,which are average for estimate the models generation performance. this is use for partition the data into the training and validating set with the 5 different iteration during k cross validation.  

here, we are using k folds with n splits for 5 folds. in short, we are splits the k folds in to the 5 folds and stored them into one variable named as kf.

In [ ]:
32561/5

6512.2

here, we are using the total number of data points and dividing by the total folds which is here is 5.  here, for the k fold cross validation we are representing the devided in to the 5 which is almost same size, while the results are shows the number of the data points in the each of the folds. here for the each of the fold serve the validate set when the model is trained and remain data in the other folds during the each iterations.
here, the k folds i division value is 6512.2.

In [ ]:
6512*4

26048

here, this calculation shows that we have used total number of data points use is 4 from the 5 folds. here, the product of the number id 6512  folds use for the training is 4 folds. this is left one fold as the validation set during the time of each iteration. the result is 26048.  



In [ ]:
i = 1
for train_set, test_set in kf.split(X=X):
    print("iteration ", i)
    print(train_set, " having :" , len(train_set))
    print(test_set, " having :" , len(test_set))
    print("-------------------------")
    i += 1

iteration  1
[ 6513  6514  6515 ... 32558 32559 32560]  having : 26048
[   0    1    2 ... 6510 6511 6512]  having : 6513
-------------------------
iteration  2
[    0     1     2 ... 32558 32559 32560]  having : 26049
[ 6513  6514  6515 ... 13022 13023 13024]  having : 6512
-------------------------
iteration  3
[    0     1     2 ... 32558 32559 32560]  having : 26049
[13025 13026 13027 ... 19534 19535 19536]  having : 6512
-------------------------
iteration  4
[    0     1     2 ... 32558 32559 32560]  having : 26049
[19537 19538 19539 ... 26046 26047 26048]  having : 6512
-------------------------
iteration  5
[    0     1     2 ... 26046 26047 26048]  having : 26049
[26049 26050 26051 ... 32558 32559 32560]  having : 6512
-------------------------



here, we are uisng for loop for iterating the loop. first its set at iterate =1 and its getting +1 after each of the iterations. we are iterating for the 5 folds of the k fold cross validation which is applied to the dataset x. here, for the each of the iteration, its print the token of the training and testing sets. here training set is containg the 80% of the data,  testing set contains the remaining 20%, this is to check that we are using the validation correctly once for all the 5 iterations.
focusing on the values there is two calculation results for having. first iteration one having is 26048 and 6512 and after iteration 2 to upto iteration 5 while the iteration having is remain same 26049 and 6512.

In [ ]:
num_cols = X.select_dtypes(include=np.number).columns
num_cols

Index(['age', ' fnlwgt', ' education-num', ' capital-gain', ' capital-loss',
       ' hours-per-week'],
      dtype='object')

here, we are selecting the numeric column from the dataframe. for selecting the columns which have numeric values only we are using x.select_dtype(include=np.number).columns  this is specify that we are choosing the column which have numberic data using np.number with selecting the data types.

focusing on the results what we got is in the index we got columns with the numeric value and columns are age, fnlwgt, education-num, capital-gain, capital-loss and hours-per-week.



In [ ]:
cat_cols = X.select_dtypes(exclude=np.number).columns
cat_cols

Index([' workclass', ' education', ' marital-status', ' occupation',
       ' relationship', ' race', ' sex', ' native-country'],
      dtype='object')

this code is totally opposite to the above. here, we are excluding the numeric data and only consider categorical data columns.

here we are using X.select_dtypes(exclude=np.number).columns which is actully excluding the columns which have numeric data types and use other data types only. stored results into the cat_cols.
focusing on the output we have workclass, education, marital-status, occupation, relationship, race sex and native-country.


In [ ]:
ct = ColumnTransformer([
    ('rob', RobustScaler(), num_cols),
    ('ohe', OneHotEncoder(handle_unknown='ignore'), cat_cols)
])

here, this code is using for transfering columns to the different subsets. here, we are using the robustscaler for handling the numerical columns , and onehotencoder is use for encoding the categorical data. for that we are using the one hot encoding with the non numeric category. here handle_unknown= ignore set if the unknown category appear then its ignored that not include for not causing any error. this is used for machine learning model for consistent preprocessing.


In [ ]:
pipe = Pipeline([
    ('ct_step', ct),
    ('model', RandomForestClassifier(n_estimators=10, random_state=0))
])

here we are creating the machine learning pipeline with contain of two steps, first is ct_step which is applied the previous define column transfer for preprocess the data. the second is model, which is the random forest classifier with the 10 decision trees and a random state 0. this pipeline is used for preprocession and creating the model.


In [ ]:
X.loc[6513:].head()

age      workclass   fnlwgt      education   education-num  \
6513   29        Private   280344      Bachelors              13   
6514   45        Private   202496      Bachelors              13   
6515   61   Self-emp-inc   134768   Some-college              10   
6516   40        Private   175686   Some-college              10   
6517   24        Private   194748        HS-grad               9   

           marital-status          occupation    relationship    race  \
6513   Married-civ-spouse               Sales         Husband   White   
6514   Married-civ-spouse   Handlers-cleaners         Husband   White   
6515   Married-civ-spouse               Sales         Husband   White   
6516             Divorced        Adm-clerical       Unmarried   White   
6517        Never-married    Transport-moving   Not-in-family   White   

          sex   capital-gain   capital-loss   hours-per-week  native-country  
6513     Male              0              0               40   United-States  
6514     Male              0              0               37   United-States  
6515     Male              0              0               40   United-States  
6516   Female              0              0               40   United-States  
6517   Female              0              0               49   United-States

this code is used for getting rows and displaying in the first five rows of the data using iloc. here we  provide the starting index which is 6513 and head() function whihc shows only  5 rows from the data and its start with the location which is we provided.  this is used for only using the useful data.


In [ ]:
scores = []
i = 1
for train_set, test_set in kf.split(X):
    pipe.fit(X.loc[train_set], y[train_set])
    sco = pipe.score(X.loc[test_set], y[test_set])
    scores.append(sco)
    print("iteration ", i)
    i += 1

iteration  1
iteration  2
iteration  3
iteration  4
iteration  5


this is use for  calculating the accuracy of the each iteration for the 5 fold. here, we are train the pipeline for the each of the iteration and also evaluate the accuracy of the test set. here the score is shows the accuracy for the eachof the iteration.

The code calculates the accuracy scores for each iteration of a 5-fold cross-validation using the defined pipeline (`pipe`). It trains the pipeline on the training set for each iteration and evaluates the accuracy on the corresponding test set. The warning messages indicate a future deprecation related to the `sparse` parameter in the `OneHotEncoder`, suggesting that the parameter name will change in future versions of scikit-learn. The output is a list (`scores`) containing the accuracy scores for each fold.

In [ ]:
np.array(scores)

array([0.84784278, 0.84520885, 0.84613022, 0.84858722, 0.85165848])

here, we are creating the array for the scores. here we are using np which is alias of numpy library for array.
here, array[0.84784278, 0.84520885, 0.84613022, 0.84858722, 0.85165848] shows the accuracy score which is calculated after each of the iteration.  this is the performance of the our random forest model. here the accuracy is lies between 0.845 to 0.851, which shows that how our dataset is consistence and how its give us high accuracy at the each iteration folds.  



In [ ]:
np.array(scores).mean()

0.8478855085142512

here, we are calculating the mean of the above accuracy scores, which shows the average performance of the random forest model after the 5 iteration of the cross validation. this is the real calculation that how well our model perform with unseen data, focusing on the mean value its 0.847 which shows our model is perfrorming better with good pridiction accuracy of the different folds.


In [ ]:
np.array(scores).std()

0.0022349531977626388

here, we are calculating the standard deviation of the above accuracy scores, which is 0.002 shows that there is variance of distribution of the model performance for the 5 iteration of cross validation. here, the standard deviation value is lower which shows that the results are more stable and constistent, which shows that model performance is stable at each of the iteration during the cross validation.


###2. Stratified KFold Cross Validation

In [ ]:
y.value_counts()

 <=50K    24720
 >50K      7841
Name:  income, dtype: int64

stratified k fold cross validation is the technique where we are distributing the data into the same class same like original data. this is use for creating the model and its good if we are using the classification and evaluating the model.
here, we are counting the y values for that we are using y,value_count().here y is the income variable which is have two classes <=50k is with 24720 and >=50k with 7841 samples. focusing on this its clearly depicts that our dataset is imbalanced where the majority of the samples are from <=50k, this is shows the distribution and highlights the technique such as stratified sample or cross validation for checking if there is bias during the model training and evaluating.


In [ ]:
7841/5

1568.2

here, we are calculating the number of total samples with 5 folds, where 7841/5 which calculate the 1568.2. here we are distributing the >=50k class evenly distribute over the 5 folds.

In [ ]:
24720/5

4944.0

here, we are calculating the number of total samples divide by the total 5 iteration where the income is <=50k where all the classes are equaly distributed for the each of the folds. here, 24720/5 is 4944.0. here, we are ensuring to have the stable data in each of the folds, which is important if we are using the stratified k fold cross validation.


In [ ]:
4944*4

19776

here, we are calculating the , 4944*4, 19776, this shows the total number of the examples in the <=50k of the class which is use for the training our model for 4 folds out of the 5 folds during the stratified k folds cross validation. here the remain 1 fold is use for performing the validation, which is give assurance to have the coverage of the both training and evaluation model.


In [ ]:
skf = StratifiedKFold(n_splits=5)

here, we are just defining the stratifiedFold with n splits of 5 and creating the object of this named as skf. this is distributed in the equal order or same class distribution just like original data.  this is deal with the imbalanced dataset, this is provide more real evaluation of the model performance during the different iterations.


In [ ]:
scores_skf = []
i = 1
for train_set, test_set in skf.split(X, y):
    pipe.fit(X.loc[train_set], y[train_set])
    sco = pipe.score(X.loc[test_set], y[test_set])
    scores_skf.append(sco)
    print("iteration ", i)
    i += 1

iteration  1
iteration  2
iteration  3
iteration  4
iteration  5


this is the stratified k fold using the skf object. here, we are  fitting our machine learning model at each of the iteration, and calculating the accuracy of the each of the iteration. here, we are using 5 iteration for fitting the model and calculating the accuracy.



In [ ]:
scores_skf

[0.8473821587594043,
 0.8432125307125307,
 0.8421375921375921,
 0.8425982800982801,
 0.8536547911547911]

here, we are getting the score of the skf which is the accuracy score for the 5 iteration. [0.8473821587594043, 0.8432125307125307, 0.8421375921375921, 0.8425982800982801, 0.8536547911547911] focusing on the score our accuracy lies in between 0.842 to 0.854, which is show the consistency of the data because of the high accuracy at the different folds. this score represent the performance of the machine learning model for the each of the iteration.
focusing on the accuracy score our model is working good because of the high accuracy.


In [ ]:
i = 1
for train_set, test_set in skf.split(X=X, y=y):
    print("iteration ", i)
    print(train_set, " having :" , len(train_set))
    print(test_set, " having :" , len(test_set))
    print()
    print("y train counts: \n", y[train_set].value_counts())
    print("y test counts: \n", y[test_set].value_counts())
    print("-------------------------")
    i += 1

iteration  1
[ 6499  6500  6512 ... 32558 32559 32560]  having : 26048
[   0    1    2 ... 6514 6515 6516]  having : 6513

y train counts: 
  <=50K    19776
 >50K      6272
Name:  income, dtype: int64
y test counts: 
  <=50K    4944
 >50K     1569
Name:  income, dtype: int64
-------------------------
iteration  2
[    0     1     2 ... 32558 32559 32560]  having : 26049
[ 6499  6500  6512 ... 13121 13123 13125]  having : 6512

y train counts: 
  <=50K    19776
 >50K      6273
Name:  income, dtype: int64
y test counts: 
  <=50K    4944
 >50K     1568
Name:  income, dtype: int64
-------------------------
iteration  3
[    0     1     2 ... 32558 32559 32560]  having : 26049
[12997 12999 13000 ... 19727 19729 19733]  having : 6512

y train counts: 
  <=50K    19776
 >50K      6273
Name:  income, dtype: int64
y test counts: 
  <=50K    4944
 >50K     1568
Name:  income, dtype: int64
-------------------------
iteration  4
[    0     1     2 ... 32558 32559 32560]  having : 26049
[19482 1948

focusing on the results its shows the 5 iteration of the straitified k fold cross validation. in each of the iteration, here  our data is splitting into two parts training and testing. at the last y count provide the satisfactory results of the income variable during the model evaluations.


In [ ]:
result_kf = cross_val_score(estimator=pipe, X=X, y=y, scoring='accuracy', cv=5)

here, we are using cross validation score for performing the cross validation of the model evaluation using the pipeline. here stratified is applied because of the default behavour of the cross val score where cv is set for the integer.


In [ ]:
result_kf

array([0.84738216, 0.84321253, 0.84213759, 0.84259828, 0.85365479])

here, we are printing the accuracy score results for kf. [0.84738216, 0.84321253, 0.84213759, 0.84259828, 0.85365479]focusing on the results its clearly visible that our model is performing well and our accuracy score is also high. accuracy score is lies in between 0.842 to 0.853. this suggest that our model is performing well.

###3. Leave one out Cross Validation

In [ ]:
start = time.time()
result_kf10 = cross_val_score(estimator=pipe, X=X, y=y, scoring='accuracy', cv=KFold(n_splits=10))
result_kf10
print("time taken: ", time.time()-start)

time taken:  30.913583040237427


Leave - one -out cross validation is use with each of the data points in the data usede as the unique test set others are use as the training data. this creates the folds as many as the data points for the each of the iteration, here model train on the all other data points.
this is show that we are using the 10 folds for split which take 30.66 seconds for complete.


In [ ]:
result_kf10

array([0.83880872, 0.85165848, 0.84981572, 0.84367322, 0.85135135,
       0.84613022, 0.84520885, 0.84797297, 0.8544226 , 0.84459459])

here, we are getting the accuracy score for the kf10, [0.83880872, 0.85165848, 0.84981572, 0.84367322, 0.85135135,0.84613022, 0.84520885, 0.84797297, 0.8544226 , 0.84459459] this is the accuracy score which shows our model is performing well. our accuracy is lies between 0.838 to 0.854. this shows that our model is performing well.

In [ ]:
start = time.time()
result_loocv = cross_val_score(estimator=pipe, X=X.head(100), y=y.head(100),
                               scoring='accuracy', cv=LeaveOneOut())
print("time taken: ", time.time()-start)

time taken:  4.312939405441284


here, we are  calcualting the time taken for performing the leave one out cross validation using the pipeline on the 100 row of the data. here we are using x with head(100) and y with head(100). here, each of the data points is taken as unique test for the each of the iteration.

here the time is 4.84 seconds taken for performing LOOCV. this is expensive and its fit the model for the each data seperately.


In [ ]:
32531/100

325.31

here, we are dividing the total number of sample with the 100 iteration and values is calculate is 325.41.

In [ ]:
325*8

2600

here, we are multiplying 325*8 and its calculate as 2600.

In [ ]:
325*8/60

43.333333333333336

here, we are calcualting the cross validation where 325*4  and divide by 60 which is 43.33.


In [ ]:
result_loocv

array([1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., 1., 1.,
       1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
       0., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 1., 1.,
       1., 1., 0., 1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 1., 0.,
       0., 1., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 0.,
       1., 0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 1., 0., 1., 1.])

here , the accuracy for  LOOCV is [1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., ...] this is the accuracy score for each of the iteration of the LOOCV. here 1.0 is correct prediction and 0.0 is incorrect prediction.

this is accuracy of the each data points.

In [ ]:
result_loocv.mean()

0.8

here, we are calcualting the mean of the accuracy score and here 0.8 is the mean of the accuracy score, which shows that our model is performing well with the data.

###4. Repeated KFold Cross Validation

In [ ]:
start = time.time()
result_rkf = cross_val_score(estimator=pipe, X=X, y=y, scoring='accuracy',
                              cv=RepeatedKFold(n_splits=5, n_repeats=5))
result_rkf
print("time taken: ", time.time()-start)

time taken:  64.08226180076599


here, we are calculating the  cross validation score and for that time taken for performing the k-fold cross validation. here, time taken is 61.48 seconds.

In [ ]:
result_rkf

array([0.84784278, 0.84689803, 0.84766585, 0.85211916, 0.84735872,
       0.84661446, 0.8519656 , 0.85657248, 0.84413391, 0.84213759,
       0.84538615, 0.85165848, 0.8416769 , 0.85165848, 0.85012285,
       0.84937817, 0.85027641, 0.84551597, 0.84566953, 0.84413391,
       0.85122064, 0.8495086 , 0.84628378, 0.85365479, 0.84920147])

this is the accuracy score array. focusing on the results our model is performing well because our accuracy lies in between 0.838 to 0.852.

###5. Performing Cross validation on a training set, and then verifying our results on a seperate testset to generalize our results.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

here, we are splitting the dataset using the train test split () along with x and y and test size is 20% with random state is 0. and we are storing that in the four variables X_train, X_test, y_train, y_test.

In [ ]:
start = time.time()
result_tts = cross_val_score(estimator=pipe, X=X_train, y=y_train,
                              scoring='accuracy', cv=KFold(n_splits=5))
print("time taken: ", time.time()-start)

time taken:  8.376204490661621


here, we are calculating the time taken to perfom the cross validation score calculation. focusing on the results its shows that time taken for this is 9.06  seconds. here, we are using the 5 k folds or 5 iteration for performing the cross validation score calculation.

In [ ]:
result_tts

array([0.85547025, 0.84702495, 0.84184261, 0.84661163, 0.84968324])

here, this is the accuracy score. which suggest that our model is performing well. focusing on the results our accuracy is lies between 0.841 to 0.855. this the accuracy for iteration of the 5 folds.

In [ ]:
pipe.fit(X_train, y_train)

Pipeline(steps=[('ct_step',
                 ColumnTransformer(transformers=[('rob', RobustScaler(),
                                                  Index(['age', ' fnlwgt', ' education-num', ' capital-gain', ' capital-loss',
       ' hours-per-week'],
      dtype='object')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore'),
                                                  Index([' workclass', ' education', ' marital-status', ' occupation',
       ' relationship', ' race', ' sex', ' native-country'],
      dtype='object'))])),
                ('model',
                 RandomForestClassifier(n_estimators=10, random_state=0))])

here, we are trying to fit our model using X_train, y_train. pipe.fit() is use for fitting the model on the training data. this process is transforming the features with intializedcolumntransfer and now training the randomforest classifire model.


In [ ]:
pipe.score(X_test, y_test)

0.8446184553968985

here, we are calculating the pip score using the testing set and calculating the accuracy of the model with the test data. here, we are getting 0.84 accuracy score, which shows that 84% of the prediction are correct with this model.
focusing on the accuracy its suggest that model performance is good.


In [ ]:
cross_val_score(estimator=pipe, X=X_train, y=y_train,scoring='accuracy', cv=KFold(n_splits=5))

array([0.85547025, 0.84702495, 0.84184261, 0.84661163, 0.84968324])

here, we are evaluating the model performance using the machine learning model pipe with k-fold cross validation with 5 iteration on the training data.[0.8555, 0.8470, 0.8418, 0.8466, 0.8497] focusing on the accuracy score this model accuracy score lies in 0.84 to 0.85 which shows that model is predicting correct results upto 84%.  

this is suggest that model is performing well with the test data.

In [ ]:
import sklearn
sorted(sklearn.metrics.SCORERS.keys())

['accuracy',
 'adjusted_mutual_info_score',
 'adjusted_rand_score',
 'average_precision',
 'balanced_accuracy',
 'completeness_score',
 'explained_variance',
 'f1',
 'f1_macro',
 'f1_micro',
 'f1_samples',
 'f1_weighted',
 'fowlkes_mallows_score',
 'homogeneity_score',
 'jaccard',
 'jaccard_macro',
 'jaccard_micro',
 'jaccard_samples',
 'jaccard_weighted',
 'matthews_corrcoef',
 'max_error',
 'mutual_info_score',
 'neg_brier_score',
 'neg_log_loss',
 'neg_mean_absolute_error',
 'neg_mean_absolute_percentage_error',
 'neg_mean_gamma_deviance',
 'neg_mean_poisson_deviance',
 'neg_mean_squared_error',
 'neg_mean_squared_log_error',
 'neg_median_absolute_error',
 'neg_negative_likelihood_ratio',
 'neg_root_mean_squared_error',
 'normalized_mutual_info_score',
 'positive_likelihood_ratio',
 'precision',
 'precision_macro',
 'precision_micro',
 'precision_samples',
 'precision_weighted',
 'r2',
 'rand_score',
 'recall',
 'recall_macro',
 'recall_micro',
 'recall_samples',
 'recall_weighted',

herem we are importing some scoring metrics for evaluating the model. for the  
classification metrics we are using accuracy, f1, precision, recall etc. regression metrics we are using negative mean squared error and negative mean absolute error.for clustering metrics we are using adjusted random score and adjusted mutual information score.


###Write your observations of the above tutorial here.**(5%)**

Here, we have use data set for income evaluation and making model and fitting model with different cross validation techniques.
data preparation:
first we are importing the libraries which is we are going to use in this tutorial. pandas , numpy, randomforestclassifier and matplotlib libraries are imported.
using pandas library we are reading and loading the file. the read_csv() function is for reading the file and also displaying the first 5 row and last five rows using head() and tail(). after that we are handling missing values using the inbuild technique of placeholder.
now, we are splitting the dataset into x and y, where x is the all columns but not drop income and y is consider income column.
cross-validation:
cross-validation is the techniques to use for evaluating the model performance using different cross validation techniques. here, we are going to use some of them such as k-fold cross-validation, straitified k fold cross validation and we are comparing the model and also using leave one out cross validation and repeated k fold cross validation and train test split cross validation for model training and evaluating.
k-fold cross validation: this is


KFold Cross-Validation:this is divide the data into the k folds and k-1 fold is used for training the model and remaining is used for testing the model with each iteration. this is repeat upto k times.

straitified cross validation: this is same like k fold but this folds have equally distributed samples or classes as the entire data. this is used with the imbalance dataset.

Leave-one-out-validation:  this is use all the data as the training data except one and testing the model using that 1 folds. eg. using 4 iteration for training the model and use 1 fold for testing the model.

Repeated k-fold cross validation: this is extends the kfold by multiple times and this is split the data randomly. this is give more realiable performance estimation because its run multiple of the times.
train-test split cross validation: this is splitting the data into two parts training and testing. this model is train in to the training set and evaluated in the testing set. this is majorly use for cross validation.

Model training and evaluation: after performing each of the operation of the cross validation now, we are training our randomforest modle and evaluating using the train test split cross validation. here, we have use pipeline for preprocessing which is columntransformer and a randomforestclassifier.

now, we are working on the performance metrics to check the performance of the model for that we have use scoring='accuracy'. here, performance is various with different cross validation techniques.
we are also calculating the time which is taken for performing the cross validation using the different techniques such as kfold, straitfiedkfold,leave-one-out, repeatedkfold and train test split.
Performance Metrics:

now, we are testing our model for that first we are fitting our model into the pipeline and use training test.
focusing on the model performance, accuracy for the model is 84-85% which is lies in between with different cross validation techniques.LOOCV is costly and time taking compared to others. repeated k fold shows some variance in the repeatation.

all in all, here, we are using investment evaluation data for checking the model performance with different types of cross validation technique. overall model performance for all the cross validation technique is arround 84 to 85% of the accuracy score, which suggest that our model is performing well.


###**Task-1(40%)**
#####Perform cross validation on the tutorial given using the dataset attached.After you are successful in executing and fixing the issues with the current version use dataset of your choice and execute it again.

###Tutorial

In [ ]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
iris=load_iris()
X=iris.data
Y=iris.target
# print("Size of Dataset {}".format(len(X)))
# logreg=LogisticRegression()
# x_train,x_test,y_train,y_test=train_test_split(X,Y,test_size=0.3,random_state=42)
# logreg.fit(x_train,y_train)
# predict=logreg.predict(x_test)
# print("Accuracy score on training set is {}".format(accuracy_score(logreg.predict(x_train),y_train)))
# print("Accuracy score on test set is {}".format(accuracy_score(predict,y_test)))

####Monte Carlo Cross Validation

In [ ]:
from sklearn.model_selection import ShuffleSplit,cross_val_score
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
import warnings
from sklearn.exceptions import ConvergenceWarning

In [ ]:
logreg=LogisticRegression(solver='sag', max_iter=1000)
shuffle_split=ShuffleSplit(test_size=0.3,train_size=0.5,n_splits=10)
warnings.filterwarnings("ignore", category=ConvergenceWarning)

In [ ]:
scores=cross_val_score(logreg,iris.data,iris.target,cv=shuffle_split)

In [ ]:
print("cross Validation scores:n {}".format(scores))
print("Average Cross Validation score :{}".format(scores.mean()))

cross Validation scores:n [0.97777778 0.97777778 0.93333333 0.97777778 0.95555556 1.
 0.97777778 0.95555556 0.97777778 0.97777778]
Average Cross Validation score :0.971111111111111


####Time Series Cross Validation

In [ ]:
import numpy as np
from sklearn.model_selection import TimeSeriesSplit
X = np.array([[1, 2], [3, 4], [1, 2], [3, 4], [1, 2], [3, 4]])
y = np.array([1, 2, 3, 4, 5, 6])
time_series = TimeSeriesSplit()
print(time_series)
for train_index, test_index in time_series.split(X):
    print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]

TimeSeriesSplit(gap=0, max_train_size=None, n_splits=5, test_size=None)
TRAIN: [0] TEST: [1]
TRAIN: [0 1] TEST: [2]
TRAIN: [0 1 2] TEST: [3]
TRAIN: [0 1 2 3] TEST: [4]
TRAIN: [0 1 2 3 4] TEST: [5]


**OWN Dataset wine**

In [ ]:
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
wine=load_wine()
X=wine.data
Y=wine.target

**Monte Carlo Cross Validation**

In [ ]:
from sklearn.model_selection import ShuffleSplit,cross_val_score
from sklearn.datasets import load_wine
from sklearn.linear_model import LogisticRegression

In [ ]:
logreg=LogisticRegression(solver='sag', max_iter=1000)
shuffle_split=ShuffleSplit(test_size=0.3,train_size=0.5,n_splits=10)
warnings.filterwarnings("ignore", category=ConvergenceWarning)

In [ ]:
scores=cross_val_score(logreg,wine.data,wine.target,cv=shuffle_split)

In [ ]:
print("cross Validation scores:n {}".format(scores))
print("Average Cross Validation score :{}".format(scores.mean()))

cross Validation scores:n [0.74074074 0.74074074 0.68518519 0.74074074 0.81481481 0.77777778
 0.74074074 0.83333333 0.81481481 0.85185185]
Average Cross Validation score :0.7740740740740741


**Time Series Cross Validation**

In [ ]:
import numpy as np
from sklearn.model_selection import TimeSeriesSplit
X = np.array([[1, 2], [3, 4], [1, 2], [3, 4], [1, 2], [3, 4]])
y = np.array([1, 2, 3, 4, 5, 6])
time_series = TimeSeriesSplit()
print(time_series)
for train_index, test_index in time_series.split(X):
    print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]

TimeSeriesSplit(gap=0, max_train_size=None, n_splits=5, test_size=None)
TRAIN: [0] TEST: [1]
TRAIN: [0 1] TEST: [2]
TRAIN: [0 1 2] TEST: [3]
TRAIN: [0 1 2 3] TEST: [4]
TRAIN: [0 1 2 3 4] TEST: [5]


###**Task-2(25%)**
###Question-2(a)  **(15)%**
####Explain the types of cross validation techniques. Also write the pros and cons of each validation technique in detail.(7 types).

Cross-validation is the technique use for evaluating the performance of the machine learning model with parting the data into the sub parts and evaluate the model with different subsets.

1. K-Fold Cross-Validation:
k fold cross validation is use in the machine learning for evaluating performance of the model by splitting the data into the k-equal size folds and train the model k times. here, every time each of the folds in used for testing the model while k-1 fold is used for training the model. this process is iterate k times.
pros:
this is more stronger model and provide accurate model performance using this method.
model provide less impact of the randomness in the data.
cons:
calculating cost is higher.
sometime its difficult for choicing the k data.

2. Holdout Validation:
this is also called as train-test split, this is use for evaluating the performance of the model. here, data is split into two parts, train and test set. here, model is train on the train set and then evaluate the model using the different test set for estimating the performance of the unknown data.
pros:
its easy and simple to use.
calculation cost is lower than other techniques.
cons:
this is selecting the random data so may be the  data splitting is too much sensitive.

3. Straitified k-fold cross validation:
this is also similar like k-fold but this technique is ensuring that divided partition for folds are must be equal and have same classes of distribution like original data.
pros:
this is use with unbalanced data.
this technique provide more accurate performance of model when its work with imbalance data and its classes distributed.
cons:
computing the performance is more expensive.

4. Leave-out-one-cross validation:
this is technique where the number of sample is set to  k. here, model train for all the data points and leave one out for testing the model or validating model.
pros:
this model is give the baised performance of the model.
small dataset is use for this model.
cons:
its expensive with large amount of the data.
results are highly variante.

5. Leave p-out cross validation:
this is same like LOOCV, only difference is p observation is not used for training  and use for validation in the each of the iteration.
pros:
this is less computation cost while comparing to the LOOCV.
cons:
its expensive, because its iterate each time with p observation with the total number of observation.

6. Time-series cross validation:
this is work with time series data, order of the data is consider here. split the data into training and validating set the way to manage the temporal order of the data.
pros:
this is use the real world scenarios and predictions is made depend on the future data.
cons:
this is only use with time series data.
training set is increase when each time the fold iterate, its difficult if its the limited data.

7. Repeated k fold cross validation:
repeat k-fold cross validation multiple times with the different random data splits.
pros:
give more real performance of the model with average the results over the multiple run.
cons:
computing of the iteration its increase.


###Question-2(b)  **(10%)**
####Provide some examples of real-world problems that require cross validation techniques?

cross-validation is important for real world machine learning for evaluate the model effects.
1. credit scoring:
this is model for predicting the credit of each of the individuals depend on the financial and personal information.
for evaluating the cross validation performance with different subset, check and ensure that its work well with the new credit scores with new applications.

2. medical diagnosis: using patient data making model to predict the  medical condition.
for checking the general performance of the model with the different patients data, here its ensure that model work well with the different cases.

3. Sentiment Analysis with NLP:
using natural language processing sentiment analysis model is created with sentiment such as positive, negative nad neutral of the text data.
for evaluating the model if its working performance is better with different domains, for ensuring the performance is well with the different text data.

4. Image classification in the autonomous vehicles:
in this model is created for detecting the object in the image which is taken by camera of the autonomous vehicels.
this model is classify the object in more accurately with different scenarios.

5. financial fraud detection:
making the model to detect the frauds in the transaction in the financial data.
this model is help to indentify the effect in the finding the fraud transaction.
use model for fetching the fraud patterns with different time periods and different transaction types.

6. prediction in the maintenance in the manufacturing:
create model to predict the instruments and equipments if its failure in the manufacturing process using the sensor data.
checking that model is work well with general manners with different types of machines and production environment, this is give the more real prediction for the preventing maintenance.

all in all, all this are the real world examples of prediction models use with the cross validation for getting better accurate performance of the model.

###**Task-3(30%)**
###Question-3(a)
Consider a dataset with 4 samples. Perform KFold, repeated KFold and LOO Cross validations.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import KFold, RepeatedKFold, LeaveOneOut, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import RobustScaler, OneHotEncoder
from sklearn.pipeline import Pipeline


wine = load_wine()

X1 = wine.data[:4]
y1 = wine.target[:4]

num_cols1 = np.arange(X1.shape[1])
cat_cols1 = []

ct1 = ColumnTransformer([
    ('rob', RobustScaler(), num_cols1),
    ('ohe', OneHotEncoder(handle_unknown='ignore'), cat_cols1)
])


model1 = RandomForestClassifier(n_estimators=10, random_state=0)


pipe1 = Pipeline([
    ('ct_step', ct1),
    ('model', model1)
])


kf1 = KFold(n_splits=2)
scores_kfold1 = cross_val_score(pipe1, X1, y1, cv=kf1)


rkf1 = RepeatedKFold(n_splits=2, n_repeats=3)
scores_rkf1 = cross_val_score(pipe1, X1, y1, cv=rkf1)


loo1 = LeaveOneOut()
scores_loo1 = cross_val_score(pipe1, X1, y1, cv=loo1)


print("KFold Cross-Validation Scores:", scores_kfold1)
print("###########")
print("Repeated KFold Cross-Validation Scores:", scores_rkf1)
print("###########")
print("Leave-One-Out Cross-Validation Scores:", scores_loo1)


KFold Cross-Validation Scores: [1. 1.]
###########
Repeated KFold Cross-Validation Scores: [1. 1. 1. 1. 1. 1.]
###########
Leave-One-Out Cross-Validation Scores: [1. 1. 1. 1.]


**K-Fold Cross Validation:**

In [ ]:
df1 = pd.read_csv('/content/Stores.csv', na_values=' ?')
df1.head()



Store_ID   Store_Area  Items_Available  Daily_Customer_Count  Store_Sales
0          1        1659             1961                   530        66490
1          2        1461             1752                   210        39820
2          3        1340             1609                   720        54010
3          4        1451             1748                   620        53730
4          5        1770             2111                   450        46620

In [ ]:
df1.isna().sum()
df1.fillna('missing', inplace=True)
df1.isna().sum()



Store_ID                0
Store_Area              0
Items_Available         0
Daily_Customer_Count    0
Store_Sales             0
dtype: int64

In [ ]:
X1 = df1.drop('Store_Sales', axis=1)
y1 = df1['Store_Sales']
kf1 = KFold(n_splits=4)

i = 1
for train_set, test_set in kf1.split(X=X1):  # Fix the variable name here
    print("iteration ", i)
    print(train_set, " having:", len(train_set))
    print(test_set, " having:", len(test_set))
    print("-------------------------")
    i += 1

num_cols1 = X1.select_dtypes(include=np.number).columns  # Fix the variable name here
cat_cols1 = X1.select_dtypes(exclude=np.number).columns  # Fix the variable name here

ct1 = ColumnTransformer([
    ('rob', RobustScaler(), num_cols1),
    ('ohe', OneHotEncoder(handle_unknown='ignore'), cat_cols1)
])

pipe1 = Pipeline([
    ('ct_step', ct1),
    ('model', RandomForestClassifier(n_estimators=10, random_state=0))
])

scores = []
i = 1
for train_set, test_set in kf1.split(X1):  # Fix the variable name here
    pipe1.fit(X1.loc[train_set], y1[train_set])
    sco = pipe1.score(X1.loc[test_set], y1[test_set])
    scores.append(sco)
    print("iteration ", i)
    i += 1

np.array(scores)
np.array(scores).mean()
np.array(scores).std()


iteration  1
[225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242
 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260
 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278
 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296
 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312 313 314
 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332
 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348 349 350
 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366 367 368
 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384 385 386
 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401 402 403 404
 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420 421 422
 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438 439 440
 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455 456 457 458
 459 460 461 462 463 464 465 466 467 4

0.0019245008972987524

**Leave one out Cross Validation:**

In [ ]:
start = time.time()
result_kf101 = cross_val_score(estimator=pipe1, X=X1, y=y1, scoring='accuracy', cv=KFold(n_splits=10))
result_kf101
print("time taken: ", time.time()-start)


time taken:  4.980360507965088


In [ ]:
result_kf101

array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.01111111, 0.        , 0.        , 0.        ])

In [ ]:
start = time.time()
result_loocv1 = cross_val_score(estimator=pipe1, X=X1.head(4), y=y1.head(4),
                               scoring='accuracy', cv=LeaveOneOut())
print("time taken: ", time.time()-start)

time taken:  0.12142229080200195


In [ ]:
start = time.time()
result_loocv = cross_val_score(estimator=pipe1, X=X1.head(4), y=y1.head(4),
                               scoring='accuracy', cv=LeaveOneOut())
print("time taken: ", time.time()-start)

time taken:  0.19169163703918457


In [ ]:
900/4


225.0

In [ ]:
225*8

1800

In [ ]:
225*8/60

30.0

In [ ]:
result_loocv
result_loocv.mean()


0.0

In [ ]:
import numpy as np
from sklearn.datasets import load_wine
from sklearn.model_selection import cross_val_score, LeaveOneOut
from sklearn.ensemble import RandomForestClassifier


wine = load_wine()
X2 = wine.data[:4]
y2 = wine.target[:4]


model2 = RandomForestClassifier(n_estimators=10, random_state=0)


loo2 = LeaveOneOut()
scores_loo2 = cross_val_score(model2, X2, y2, cv=loo2)


print("Leave-One-Out Cross-Validation Scores:", scores_loo2)
print("Mean Accuracy:", np.mean(scores_loo2))


Leave-One-Out Cross-Validation Scores: [1. 1. 1. 1.]
Mean Accuracy: 1.0


**Repeated KFold Cross Validation:**

In [ ]:
start = time.time()
result_rkf1 = cross_val_score(estimator=pipe1, X=X1, y=y1, scoring='accuracy',
                              cv=RepeatedKFold(n_splits=4, n_repeats=4))
result_rkf1
print("time taken: ", time.time()-start)

time taken:  5.080450773239136


In [ ]:
result_rkf1

array([0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.00444444,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        ])

In [ ]:
start = time.time()
result_rkf1 = cross_val_score(estimator=pipe1, X=X1, y=y1, scoring='accuracy',
                              cv=RepeatedKFold(n_splits=4, n_repeats=4))
print("time taken: ", time.time() - start)
print("Repeated K-Fold Cross Validation Scores:", result_rkf1)


time taken:  5.095754861831665
Repeated K-Fold Cross Validation Scores: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import RepeatedKFold, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import RobustScaler, OneHotEncoder
from sklearn.pipeline import Pipeline


wine = load_wine()
X3 = wine.data[:4]
y3 = wine.target[:4]


num_cols3 = pd.DataFrame(X3).select_dtypes(include=np.number).columns
cat_cols3 = pd.DataFrame(X3).select_dtypes(exclude=np.number).columns


ct3 = ColumnTransformer([
    ('rob', RobustScaler(), num_cols3),
    ('ohe', OneHotEncoder(handle_unknown='ignore'), cat_cols3)
])


model3 = RandomForestClassifier(n_estimators=10, random_state=0)


pipe3 = Pipeline([
    ('ct_step', ct3),
    ('model', model3)
])


rkf3 = RepeatedKFold(n_splits=2, n_repeats=2)
scores_rkf3 = cross_val_score(pipe3, X3, y3, cv=rkf3)


print("Repeated K-Fold Cross-Validation Scores:", scores_rkf3)


Repeated K-Fold Cross-Validation Scores: [1. 1. 1. 1.]


###Question-3(b)
Consider a dataset with 6 samples and perform 3 split time-series cross validation

In [ ]:
#Write your code here
import numpy as np
from sklearn.model_selection import TimeSeriesSplit


data = np.array([1, 2, 3, 4, 5, 6])


tscv = TimeSeriesSplit(n_splits=3)

for train_index, test_index in tscv.split(data):
    print("Train:", data[train_index], "Test:", data[test_index])


Train: [1 2 3] Test: [4]
Train: [1 2 3 4] Test: [5]
Train: [1 2 3 4 5] Test: [6]


In [ ]:
import numpy as np
from sklearn.model_selection import TimeSeriesSplit

X = np.array([[1, 2], [3, 4], [1, 2], [3, 4], [1, 2], [3, 4]])
y = np.array([1, 2, 3, 4, 5, 6])

tscv = TimeSeriesSplit()
print(tscv)

for train_index, test_index in tscv.split(X):
    print("TRAIN:", train_index, "TEST:", test_index)
    X_train, X_test = X[train_index], X[test_index]
    y_train, y_test = y[train_index], y[test_index]


TimeSeriesSplit(gap=0, max_train_size=None, n_splits=5, test_size=None)
TRAIN: [0] TEST: [1]
TRAIN: [0 1] TEST: [2]
TRAIN: [0 1 2] TEST: [3]
TRAIN: [0 1 2 3] TEST: [4]
TRAIN: [0 1 2 3 4] TEST: [5]
